File: TheQuizApp/app/src/main/java/com/example/thequizapp/MainActivity.java
package com.example.thequizapp;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.RadioButton;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.example.thequizapp.databinding.ActivityMainBinding;
import com.example.thequizapp.model.Question;
import com.example.thequizapp.model.QuestionList;
import com.example.thequizapp.viewmodel.QuizViewModel;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    ActivityMainBinding binding;
    QuizViewModel quizViewModel;
    List<Question> questionList;

    static int result = 0;
    static int totalQuestions = 0;
    int i = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        Log.v("MyActivity", "start" );

        binding = DataBindingUtil.setContentView(
                this,
                R.layout.activity_main
        );

        Log.v("MyActivity", "DataBinding OK");
        result = 0;
        totalQuestions = 0;

        quizViewModel = new ViewModelProvider(this).get(QuizViewModel.class);
        Log.v("MyActivity", "QuizModel OK" );

        DisplayFirstQuestion();

        binding.btnNext.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DisplayNextQuestion();
            }
        });
    }

    public void DisplayFirstQuestion() {
        quizViewModel.getQuestionListLiveData().observe(
                this,
                new Observer<QuestionList>() {

                    @Override
                    public void onChanged(QuestionList questions) {
                        questionList = questions;
                        binding.txtQuestion.setText("Question 1: " + questions.get(0).getQuestion());
                        binding.radio1.setText(questions.get(0).getOption1());
                        binding.radio2.setText(questions.get(0).getOption2());
                        binding.radio3.setText(questions.get(0).getOption3());
                        binding.radio4.setText(questions.get(0).getOption4());
                    }
                }
        );
    }

    public void DisplayNextQuestion() {

        if(binding.btnNext.getText().equals("Finish")) {
            Intent i = new Intent(MainActivity.this, ResultsActivity.class);
            startActivity(i);
            finish();
        }

        int selectedOption = binding.radioGroup.getCheckedRadioButtonId();
        Log.v("MyActivity", "RadioId=" + selectedOption);
        if(selectedOption != -1) {
            Log.v("MyActivity", "result=" + result);
            RadioButton radioButton = findViewById(selectedOption);

            if((questionList.size() - i) > 0) {
                totalQuestions = questionList.size();
                Log.v("MyActivity", "radioTxt=" + radioButton.getText().toString());
                Log.v("MyActivity", "CorrectAns=" + questionList.get(i).getCorrectOption());
                if(radioButton.getText().toString().equals(
                        questionList.get(i).getCorrectOption()
                )) {
                    result++;
                    Log.v("MyActivity", "result=" + result);
                    binding.txtResult.setText(
                            "Correct Answers: " + result
                    );
                }

                if(i == 0) {
                    i++;
                }

                binding.txtQuestion.setText("Question " + (i + 1) + " : " +
                        questionList.get(i).getQuestion());

                binding.radio1.setText(questionList.get(i).getOption1());
                binding.radio2.setText(questionList.get(i).getOption2());
                binding.radio3.setText(questionList.get(i).getOption3());
                binding.radio4.setText(questionList.get(i).getOption4());

                if(i == (questionList.size() - 1)) {
                    binding.btnNext.setText("Finish");
                }

                binding.radioGroup.clearCheck();
                i++;
            } else {
                if(radioButton.getText().toString().equals(
                        questionList.get(i - 1).getCorrectOption()
                )) {
                    result++;
                    Log.v("MyActivity", "result=" + result);
                    binding.txtResult.setText("Correct Answers: " + result);
                }
            }
        } else {
            Toast.makeText(
                    this,
                    "You need to make a selection",
                    Toast.LENGTH_SHORT).show();
        }
    }
}

File: KotlinQuizApp/app/src/main/java/com/example/kotlinquizapp/view/MainActivity.kt
package com.example.kotlinquizapp.view

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.RadioButton
import android.widget.TextView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.example.kotlinquizapp.R
import com.example.kotlinquizapp.databinding.ActivityMainBinding
import com.example.kotlinquizapp.model.Question
import com.example.kotlinquizapp.viewmodel.QuizViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    lateinit var binding: ActivityMainBinding
    lateinit var quizViewModel: QuizViewModel
    lateinit var questionsList: List<Question>

    companion object {
        var result = 0
        var totalQuestions = 0
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        Log.i("TAGY", "passou")
//        setContentView(R.layout.activity_main)
//        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
//            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
//            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
//            insets
//        }

        Log.i("TAGY", "passou")

        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        // Reseting the scores
        result = 0
        totalQuestions = 0

        // Getting the response
        quizViewModel = ViewModelProvider(this)
            .get(QuizViewModel::class.java)

        // Display the first question
        GlobalScope.launch(Dispatchers.Main) {

            quizViewModel.getQuestionsFromLiveData().observe(this@MainActivity, Observer {
                if(it.size > 0) {
                    questionsList = it
                    Log.i("TAGY", "This is the 1st question: ${questionsList[0]}")

                    binding.apply {
                        txtQuestion.text = questionsList!![0].question
                        radio1.text = questionsList!![0].option1
                        radio2.text = questionsList!![0].option2
                        radio3.text = questionsList!![0].option3
                        radio4.text = questionsList!![0].option4
                    }
                }
            })

        }

        // Adding functionality no next button
        var i = 1
        binding.apply {
            btnNext.setOnClickListener(View.OnClickListener {
                val selectedOption = radioGroup?.checkedRadioButtonId


                if(selectedOption != -1) {
                    val radButton = findViewById<View>(selectedOption!!) as RadioButton
//

                        questionsList.let {
                        if (i < it.size!!) {
                            totalQuestions = it.size
                            if(radButton.text.toString().equals(it[i-1].correct_option)) {
                                result++
                                txtResult?.text = "Correct Answer: $result"
                            }

                            // Displaying the next Questions
                            txtQuestion.text = "Question ${i+1}: " + questionsList[i].question
                            radio1.text = it[i].option1
                            radio2.text = it[i].option2
                            radio3.text = it[i].option3
                            radio4.text = it[i].option4

                            // Checking if it is the last question
                            if(i == it.size!!.minus(1)) {
                                btnNext.text = "FINISH"
                            }

                            radioGroup?.clearCheck()
                            i++
                        } else {
                            if(radButton.text.toString().equals(it[i-1].correct_option)) {
                                result++
                                txtResult?.text = "Correct Answer: $result"
                            } else {

                            }

                            val intent = Intent(this@MainActivity, ResultActivity::class.java)
                            startActivity(intent)
                            finish()
                        }
                    }
                } else {
                    Toast.makeText(this@MainActivity, "Please select on option",
                        Toast.LENGTH_LONG).show()
                }

            })
        }


    }
}